### PR TITLE
allow metadata to be passed to embedded iframes

### PR DESCRIPTION
**Pull request by tysonbattistella**

### CODE REVIEW

File: src/components/IFrame.js [hunk 1]
```diff
@@ -54,6 +55,18 @@ export default class IFrame extends React.Component<Props, State> {
     window.removeEventListener('message', this.handleMessage);
   }
 
+  /*
+   * Ensure that the metadata we are supplying to the embedded IFrame is properly URI encoded
+   */
+  encodeMetadata = (metadata: Object): string => {
+    try {
+        return encodeURI(JSON.stringify(metadata));
```
Comment (pshc) [inline src/components/IFrame.js]: `encodeURIComponent` for query params, methinks

File: src/components/IFrame.js [hunk 2]
```diff
@@ -100,17 +113,18 @@ export default class IFrame extends React.Component<Props, State> {
...
     if(!widgetType){
       throw new Error('must supply either checkoutId or chargeId prop');
     }
 
     const id = checkoutId ? checkoutId : chargeId;
+    const encodedMetadata = metadata ? this.encodeMetadata(metadata) : '';
 
     // $FlowFixMe Cannot call encodeURI with id bound to uri because undefined is incompatible with string.
-    const src = `${this.origin}/embed/${widgetType}/${encodeURI(id)}?origin=${encodeURI(this.hostName)}&version=${encodeURI(VERSION)}&buttonId=${this.uuid}`;
+    const src = `${this.origin}/embed/${widgetType}/${encodeURI(id)}?origin=${encodeURI(this.hostName)}&version=${encodeURI(VERSION)}&buttonId=${this.uuid}&custom=${encodedMetadata}`;
```
Comment (pshc) [inline src/components/IFrame.js]: might be worth stealin the query encoding function i wrote for checkout, or similar

```
    function encodeURIParams(params) {
      let encoded = []
      let quote = window.encodeURIComponent
      for (let key in params) {
        if (params.hasOwnProperty(key)) {
          encoded.push(quote(key) + '=' + quote(params[key]))
        }
      }
      return encoded.join('&')
    }
```

File: src/components/IFrame.js [hunk 3]
```diff
@@ -100,17 +113,18 @@ export default class IFrame extends React.Component<Props, State> {
...
 
     const widgetType = checkoutId ? 'checkout' : chargeId ? 'charges' : null;
     if(!widgetType){
       throw new Error('must supply either checkoutId or chargeId prop');
     }
 
     const id = checkoutId ? checkoutId : chargeId;
+    const encodedMetadata = metadata ? this.encodeMetadata(metadata) : '';
 
     // $FlowFixMe Cannot call encodeURI with id bound to uri because undefined is incompatible with string.
```
Comment (pshc) [inline src/components/IFrame.js]: can fix this $flowfixme by rewriting the above logic... maybe:

```js
let widgetType
let id
if (checkoutId) {
  id = checkoutId;
  widgetType = 'checkout';
} else if (chargeId) {
  id = chargeId;
  widgetType = 'charges';
} else {
  throw new Error(...);
}
```

File: src/components/__snapshots__/IFrame.spec.js.snap [hunk 4]
```diff
@@ -15,7 +15,7 @@ exports[`it renders and matches its snapshot 1`] = `
       className="coinbase-commerce-iframe"
       frameBorder="no"
       scrolling="no"
-      src="https://commerce.coinbase.com/embed/checkout/aaaa?origin=http://localhost&version=1&buttonId=undefined"
+      src="https://commerce.coinbase.com/embed/checkout/aaaa?origin=http://localhost&version=1&buttonId=undefined&custom="
```
Comment (pshc) [inline src/components/__snapshots__/IFrame.spec.js.snap]: `buttonId=undefined` :seemslegit:

File: src/index.js [hunk 5]
```diff
@@ -9,6 +9,7 @@ type Props = {
   children?: React.Node,
   checkoutId?: string,
   chargeId?: string,
+  metadata?: Object,
```
Comment (pshc) [inline src/index.js]: a more precise type would be `{[string]: string}` for what it's worth

File: src/components/__snapshots__/IFrame.spec.js.snap (hunk 4 above)
Comment (tysonbattistella) [reply src/components/__snapshots__/IFrame.spec.js.snap]: uuid utils are mocked so snapshots don't have different ids each time their tests are ran 🙃

File: src/components/IFrame.js [hunk 6]
```diff
@@ -54,6 +57,54 @@ export default class IFrame extends React.Component<Props, State> {
...
+  /*
+   * Extract a string-representation of the supplied metadata
+   */
+  encodeMetadata = (metadata: Object): string => {
+    try {
+        return JSON.stringify(metadata);
+    } catch (e) {
+      console.error(e);
+      return ''
+    }
```
Comment (pshc) [inline src/components/IFrame.js]: i feel like we shouldn't try-catch this... better to fail loudly than quietly omit metadata but succeed

File: readme.md [hunk 7]
```diff
@@ -55,6 +55,7 @@ In addition to the regular `button` props, this component accepts some custom pr
 | `onChargeSuccess` | nil     | no                    | `(MessageData)=>void` |
 | `onChargeFailure` | nil     | no                    | `(MessageData)=>void` |
 | `onModalClosed`   | nil     | no                    | `()=>void`            |
+| `metadata`        | nil     | no                    | `Object`              |
```
[the remark this answers is not in the document]
Comment (pshc) [reply readme.md]: wait if we're stuffing all of this metadata into just the `custom` key... ugh. it's kinda gross.

could we just have `metadata` accept a string instead?

File: src/components/IFrame.js [hunk 8]
```diff
@@ -1,4 +1,6 @@
 // @flow
+declare var VERSION: string;
```
[the remark this answers is not in the document]
Comment (tysonbattistella) [reply src/components/IFrame.js]: VERSION is injected by webpack so we can keep track of which versions are out in the wild. See  [here](https://github.com/coinbase/react-coinbase-commerce/blob/4f5cae6cd0c97c32441c56b04d115f5e3bb61b7e/webpack.prod.config.js#L61)

[the remark this answers is not in the document]
Comment (pshc) [reply src/components/IFrame.js]: neato!

File: src/components/IFrame.js [hunk 9]
```diff
@@ -50,16 +53,57 @@ export default class IFrame extends React.Component<Props, State> {
...
+      console.error('Received customMetadata not of "string" type. Ignoring.');
+    } else if (customMetadata) {
+      custom = customMetadata
+    }
+    const params = {
+      origin: this.hostName,
+      version: VERSION,
+      buttonId: this.uuid,
+      custom
+    };
```
Comment (pshc) [inline src/components/IFrame.js]: it doesn't really matter, but i would find it cleaner to not send the `custom=` query param when it's missing

File: src/components/IFrame.js [hunk 10]
```diff
@@ -25,7 +28,7 @@ export default class IFrame extends React.Component<Props, State> {
   uuid: string;
   hostName: string;
 
-  constructor(props:Props){
+  constructor(props: Props) {
     super(props);
 
     this.origin = 'https://commerce.coinbase.com';
```
Comment (pshc) [inline src/components/IFrame.js]: out of scope, but this naming is confusing. hostName ought to be origin and origin ought to be... idk commerceRoot or something

Comment (pshc) [reply src/components/IFrame.js]: whatever, bikeshed at this point

File: readme.md [hunk 11]
```diff
@@ -55,6 +55,7 @@ In addition to the regular `button` props, this component accepts some custom pr
 | `onChargeSuccess` | nil     | no                    | `(MessageData)=>void` |
 | `onChargeFailure` | nil     | no                    | `(MessageData)=>void` |
 | `onModalClosed`   | nil     | no                    | `()=>void`            |
+| `cutomMetadata`   | nil     | no                    | `string`              |
```
Comment (pshc) [inline readme.md]: typo!

File: src/index.js [hunk 12]
```diff
@@ -60,14 +61,14 @@ class CoinbaseCommerceButton extends React.Component<Props, State>{
...
-    const {onLoad, onChargeSuccess, onChargeFailure, checkoutId, chargeId} = this.props;
+    const {onLoad, onChargeSuccess, onChargeFailure, checkoutId, chargeId, customMetadata} = this.props;
     const iFrameProps = {onLoad, onChargeSuccess, onChargeFailure, checkoutId, chargeId};
     const buttonProps = getButtonProps(this.props);
     return (
       <React.Fragment>
         <Button {...buttonProps} onClick={this.handleButtonClick.bind(this)}/>
         {showModal && (
-          <IFrame {...iFrameProps} onModalClose={this.handleModalClose.bind(this)} onError={this.handleError} />
+          <IFrame {...iFrameProps} onModalClose={this.handleModalClose.bind(this)} onError={this.handleError} customMetadata={customMetadata}/>
```
Comment (pshc) [inline src/index.js]: we bind `handleModalClose` and friends to `this` up to three times, phew

`handleModalClose = () => {` already does enough binding, no need for the additional `.bind` calls

File: readme.md (hunk 11 above)
Comment (tysonbattistella) [reply readme.md]: 😞

File: src/components/IFrame.js [hunk 13]
```diff
@@ -42,16 +51,62 @@ export default class IFrame extends React.Component<Props, State> {
...
+      console.error('Received customMetadata not of "string" type. Ignoring.');
+    } else if (customMetadata) {
+      custom = customMetadata
+    }
+
+    if (custom){
+      params.custom = custom
+    }
+
+    return `${this.origin}/embed/${widgetType}/${encodeURI(id)}?${encodeURIParams(params)}`;
```
Comment (pshc) [inline src/components/IFrame.js]: nit: i believe that `encodeURI` is more correct as `encodeURIComponent` but I'm not entirely sure. `encodeURI` doesn't encode `?`s or `&`s so a bogus `id` could mess with this URI

**Pull request merged**